### PR TITLE
Block Editor: Add settings to enable/disable auto anchor generation

### LINKED ISSUE #38171: Enable & disable autogenerated anchors in heading blocks
### Description

I want to remove the automatic html anchor for the heading block. There are cases where IDs are attached to heading blocks with hooks and plugins.

I also agree with the comments (https://github.com/WordPress/gutenberg/pull/30825#issuecomment-959051708).

Please consider that the html anchor can be enabled / disabled as an option or a filter.

### Step-by-step reproduction instructions

1. Add Heading block
2. Type any text
3. Open Advance setting
4. Delete HTML anchor（hold down delete key）

### Screenshots, screen recording, code snippet

https://user-images.githubusercontent.com/15373114/150731430-e6b467ad-930d-4026-9831-3c47f1e3cb77.mov

### Environment info

WP RC-3
Gutenberg 12.4.1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

### PR DESCRIPTION
**Pull request by Mamaduka**
## Description
Resolves #38171.

This PR introduces new block editor settings - `__experimentalGenerateAnchors` and makes automatic anchor generation opt-in.

**Why use block editor settings instead of Preferences?**

Most of the Preferences are saved in the `edit-post` packages store. As we don't want to make the `block-library` package dependent on `edit-post` this was the most viable option.

## Testing Instructions
1. Open a Post or Page.
2. Add multiple headings.
3. Check that anchor aren't automatically generated.
4. Enable feature.
5. Refresh the editor.
6. Confirm that anchors are generated.
7. They update as you change heading contents.

```php
add_filter(
	'block_editor_settings_all',
	static function( $settings ) {
		$settings['__experimentalGenerateAnchors'] = true;
		return $settings;
	}
);
```

## Types of changes
Enhancement

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] I've tested my changes with keyboard and screen readers. 
- [x] My code has proper inline documentation. 
- [x] I've included developer documentation if appropriate. 
- [ ] I've updated all React Native files affected by any refactorings/renamings in this PR (please manually search all `*.native.js` files for terms that need renaming or removal). 
- [x] I've updated related schemas if appropriate.

### CODE REVIEW

**Comment by Mamaduka** [comment]
Do we want to enable this feature in tests, or should I re-generate the fixtures?

**Comment by aristath** [comment]
I'm still unsure about whether this should be opt-in or opt-out 🤔 
Other than that, the code looks good and works as expected

**Comment by Mamaduka** [comment]
Considering feedback, we might want to make this opt-in in 5.9.x and then opt-out when UX is more polished.

@youknowriad, @jasmussen, @priethor What do you think?

**Comment by jasmussen** [comment]
Without a UI, it seems safest to start conservative, opt-in by default, and then we can make it opt-out if/when a UI surfaces for it. 

I want to note that's not at all a strong opinion and I'll defer to someone who has one.

**Comment by aristath** [comment]
I think that ideally, it would be opt-out for latin languages (with some basic transliteration to remove accents etc), and opt-in for non-latin languages like Mandarin, Greek, Arabic etc where we've seen most issues.
That being said, the easiest way for now would indeed be to make it opt-in

**Comment by Mamaduka** [comment]
> That being said, the easiest way for now would indeed be to make it opt-in.

I think this is the most viable option to be backported into a minor release. We can work on UI and smart default for 6.0.

**Review by aristath — APPROVED** [review]
Following the discussion in the comments above, I'll go ahead and pre-approve this (pending re-generation of the fixtures to fix failing tests)

**Comment by Mamaduka** [comment]
Thanks for the feedback, folks. Let's try to polish UI/UX for this feature before 6.0.

I'm going to merge this PR and backport it to a minor release.

**Comment by shuttermuse** [comment]
Ok so you have gone to opt-out by default. But where do I go to opt in? A feature I liked has been disabled, and I have looked in all the seemingly obvious places to re-enable it.

**Comment by Mamaduka** [comment]
Hi, @shuttermuse

Currently, there's no UI to enable the feature back, but we'll try to ship this in a future version of WordPress.

Meanwhile, to enable it, you'll need to place the following snippet in `functions.php` or any custom plugin you use for site functionality code.

```php
add_filter(
	'block_editor_settings_all',
	static function( $settings ) {
		$settings['__experimentalGenerateAnchors'] = true;
		return $settings;
	}
);
```

**Comment by shuttermuse** [comment]
@Mamaduka Thanks for the code snippet! 

I was tearing my hair out trying to find the UI option.

**Comment by mtias** [comment]
For future reference: it also seems like it should be turned on automatically in the context of a post that uses the table of contents block.

**Comment by tp-cast-ai** [comment]
For anyone googling how to turn this back on - `__experimentalGenerateAnchors ` was renamed to `generateAnchors` with no notice, nor UI being in place to turn it back on :)

```php
add_filter( 'block_editor_settings_all', function( $editor_settings, $editor_context ) {
	$editor_settings['generateAnchors'] = true;
	return $editor_settings;
}, 10, 2 );
```

**Comment by Membele** [comment]
> For anyone googling how to turn this back on - `__experimentalGenerateAnchors ` was renamed to `generateAnchors` with no notice, nor UI being in place to turn it back on :)
> 
> ```
> add_filter( 'block_editor_settings_all', function( $editor_settings, $editor_context ) {
> 	$editor_settings['generateAnchors'] = true;
> 	return $editor_settings;
> }, 10, 2 );
> ```

Thx!

**Pull request merged**
